src/lj_trace.c

```diff
@@ -47,6 +49,42 @@ void lj_trace_err_info(jit_State *J, TraceError e)
   lj_err_throw(J->L, LUA_ERRRUN);
 }
 
+/* -- Hotcount decay ------------------------------------------------------ */
+
+/* We reset all hotcounts every second. This is a rough way to establish a
+** relation with elapsed time so that hotcounts provide a measure of frequency.
+**
+** The concrete goal is to ensure that the JIT will trace code that becomes hot
+** over a short duration, but not code that becomes hot over, say, the course
+** of an hour.
+**
+** The "one second" constant is certainly tunable.
+** */
+
+static inline uint64_t gettime_ns (void)
+{
+  struct timespec ts;
+  clock_gettime(CLOCK_MONOTONIC, &ts);
+  return ts.tv_sec * 1000000000LL + ts.tv_nsec;
+}
+
+/* Timestamp (ns) of last hotcount reset. */
+static uint64_t hotcount_decay_ts;
+
+/* Decay hotcounts every second. */
+int hotcount_decay (jit_State *J)
+{
+  uint64_t ts = gettime_ns();
+  int decay = (ts - hotcount_decay_ts) > 1000000000LL; /* 1s elapsed? */
```

Comment:
The time that pass is usualy always positive but there is a case where it is not the case.

Consider the system clock at the value 1234567890
The admin or a system service sync the clock with a time server and set the date 1hour later.
The next system clock call will return a value near `1234564290`.

With your current code, the `(ts - hotcount_decay_ts)`  will be negative, isn't it ?
Then the hotcounts will not be reseted during the drift value (one hour in my sample).

It could be easily detected and reseted with something like :
```
int decay = (ts - hotcount_decay_ts) > 1000000000LL || (ts - hotcount_decay_ts) < 0LL;
```
I let you consider the correct behavior of this case.

I'm sysadmin, I see often crash and strange behavior just after a time sync, I'm aware of this kind of issue.

Regards,
A friendly time traveler

Reply:
>  The admin or a system service sync the clock with a time server and set the date 1hour later.

It’s using `CLOCK_MONOTONIC` so this shouldn’t affect us here, i.e. the timestamp will never decrease.

Reply:
Ok perfect ! :D